File: graphs/js/minIsland.js
const grid = [
    ['w','l','w','w','w'],
    ['w','l','w','w','w'],
    ['w','w','w','l','w'],
    ['w','w','l','l','w'],
    ['l','w','w','l','l'],
    ['l','l','w','w','w']
]

const minIsland = () =>{
    const visited = new Set()

    let minSize = Infinity;
    const islandSize = {}

    for(let i=0;i<grid.length;i++){
        for(let j=0;j<grid[i].length;j++){
            if(grid[i][j]==='l'){
                const size = explore(i,j,visited)
                if(size > 0){
                minSize = size < minSize ? size: minSize
                if(!islandSize[i+','+j]){
                    islandSize[i+','+j] = size
                }

                }
            }
        }
    }
    return [minSize,islandSize]
}

const explore = (i,j,visited) => {

    const rb = 0 <= i && i < grid.length
    const cb = 0 <= j && j < grid[0].length

    if(!rb || !cb || grid[i][j] === 'w') return 0
    const pos = i + ',' + j

    if(visited.has(pos)) return 0;

    visited.add(pos)

    let size = 1

    size += explore(i+1,j,visited)
    size += explore(i-1,j,visited)
    size += explore(i,j+1,visited)
    size += explore(i,j-1,visited)

    return size


}

const [minSize,islandSize] = minIsland()
console.log("minSize: ", minSize)
console.log("IslandsSize: ", islandSize)
const getMinSize = (islandSize) => {
    let Mkey, Mval, min= Infinity

    for (const [key, value] of Object.entries(islandSize)){
        if( value < min){
            Mkey = key
            Mval = value
            min = value
        }
    }
    return [Mkey, Mval]
}
const [Mkey, Mval] = getMinSize(islandSize)

console.log("Component containing node "+Mkey+ " has smallest island size of "+Mval)

File: graphs/java/MinIsland.java
package linear_concepts.graphs.java;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

public class MinIsland {

     public static List<List<String>> grid = new ArrayList<List<String>>();
    public static HashSet<String> visited = new HashSet<String>();

    public static void main(String[] args) {
        grid.add(Arrays.asList("w","l","w","w","w"));
        grid.add(Arrays.asList("w","l","w","w","w"));
        grid.add(Arrays.asList("w","w","w","l","w"));
        grid.add(Arrays.asList("w","w","l","l","w"));
        grid.add(Arrays.asList("l","w","w","l","l"));
        grid.add(Arrays.asList("l","l","w","w","w"));

        System.out.print(minIsland());
    }

    public static int minIsland(){

        int total_size = Integer.MAX_VALUE;

        HashMap<String,Integer> islandSize = new HashMap<String, Integer>();

        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid.get(i).size();j++){
                if(grid.get(i).get(j).equals("l")){
                    int size = explore(i,j);
                    if(size > 0 ){
                        total_size = size < total_size ? size : total_size;
                        //create dictionary of sizes
                         String pos = Integer.toString(i) + "," + Integer.toString(j);
                        if(!islandSize.containsKey(pos)){
                            islandSize.put(pos, size);
                        }
                    }
                }
            }
        }

        System.out.println(islandSize);

        return total_size;
    }

    public static int explore(int i, int j){

        boolean rb = 0 <= i && i < grid.size();
        boolean cb = 0 <= j && j < grid.get(0).size();

        if( !rb || !cb || grid.get(i).get(j).equals("w")) return 0;

        String pos = Integer.toString(i) + "," + Integer.toString(j);

        if(visited.contains(pos)) return 0;

        visited.add(pos);

        int size = 1;

        size += explore(i+1, j);
        size += explore(i-1, j);
        size += explore(i, j+1);
        size += explore(i+1, j-1);

        return size;
    }
    
}
